docs(examples): Add examples readme with gifs

joshka opened the pull request:
This commit adds a readme to the examples directory with gifs of each
example. This should make it easier to see what each example does
without having to run it.

I modified the examples to fit better in the gifs. Mostly this was just
removing the margins, but for the block example I cleaned up the code a
bit to make it more readable.

NOTE!!!!

There are some problems with rendering over ssh, which are not yet resolved. These are noted in the README file. It's possible that these problems are related to https://github.com/tui-rs-revival/ratatui/issues/298.

There is a github action that can automatically run VHS as part of a CI process - which would be possibly useful to keep things up to date. https://github.com/charmbracelet/vhs-action

orhun commented:
Hey, what else is needed to be done here? Do you want me to update the workflow to add `vhs-action` maybe?

joshka commented:
Note the TODOs in the doc. I haven't had a chance to dig into why some of the problems are happening (our side vs vhs side)

orhun commented on examples/README.md:
```diff
@@ -0,0 +1,189 @@
+
+## Layout ([layout.rs](./layout.rs))
+
+TODO Layout renders poorly
```
FWIW, setting the `FontSize` to `15` or updating the constraints to 20-60-20 fixes this.

joshka replied:
yep - will do

joshka replied:
Updated the example to be even nicer.

orhun commented on examples/README.md:
```diff
@@ -0,0 +1,189 @@
+
+## Table ([table.rs](./table.rs))
+
+TODO Table Panics!
```
I couldn't reproduce this.

joshka replied:
It's likely a size constraint thing - I haven't dug in to this yet.

joshka replied:
Repro was due to the width of the space that it was rendering in, and the interaction with Length and Min Constraints. If the constraints can't be satisfied then we panic.

orhun commented on examples/README.md:
```diff
@@ -0,0 +1,189 @@
+
+## Gauge ([gauge.rs](./gauge.rs))
+
+TODO Gauges render poorly on a tty
```
Seems fine to me, only thing is the last gauge is missing borders but can be fixed via:

```diff
diff --git a/examples/gauge.rs b/examples/gauge.rs
index 5423ca7..26773d6 100644
--- a/examples/gauge.rs
+++ b/examples/gauge.rs
@@ -154,7 +154,7 @@ fn ui<B: Backend>(f: &mut Frame<B>, app: &App) {
 
     let label = format!("{}/100", app.progress4);
     let gauge = Gauge::default()
-        .block(Block::default().title("Gauge4"))
+        .block(Block::default().title("Gauge4").borders(Borders::ALL))
         .gauge_style(
             Style::default()
                 .fg(Color::Cyan)

```

joshka replied:
From my terminal:
<img width="816" alt="image" src="https://github.com/tui-rs-revival/ratatui/assets/381361/45a35ad0-82ff-4895-92a0-1bf04341f5f7">

From the image:
![image](https://github.com/tui-rs-revival/ratatui/assets/381361/fe03a2f4-a2a4-4f2c-860c-76e115eea506)

joshka replied:
Added the border to this.

joshka commented:
In Block.rs, the entire block background should be green and the top two blocks should render similar widths:
<img width="814" alt="image" src="https://github.com/tui-rs-revival/ratatui/assets/381361/34b0236f-2d7d-4624-8686-1798144808cd">
vs 
![image](https://github.com/tui-rs-revival/ratatui/assets/381361/afaf2859-9055-490f-bfa5-a1ede93cbdac)

Popup has an issue rendering the background:
Terminal:
<img width="813" alt="image" src="https://github.com/tui-rs-revival/ratatui/assets/381361/adfbca0c-59ef-430d-a251-03793e1226aa">
VHS:
![image](https://github.com/tui-rs-revival/ratatui/assets/381361/2cce75e6-2d57-4366-a7cc-c895755b2f3e)


Perhaps it would be sufficient to just note that there are rendering issues with VHS (or rewrite the examples to avoid the issues). @orhun given that you are not seeing the same issues, It's possible that there's some sort of difference in rendering on a mac vs on a linux machine due to tty differences and installed fonts - so I'll check there to compare.

orhun commented:
Yeah, it is likely that those problems are mostly about `vhs` itself so nothing that would be a good idea.

joshka commented:
Fixed block, layout and table by changing the examples a bit and updated the README with the new gifs.
Direct link to [README](https://github.com/tui-rs-revival/ratatui/blob/e2acc79036b8fa58458effcb500c9d989402e153/examples/README.md)

![image](https://github.com/tui-rs-revival/ratatui/assets/381361/2155e9e7-aeaa-4b3e-a478-d3e996563920)

![image](https://github.com/tui-rs-revival/ratatui/assets/381361/d7ac045c-b429-4cad-b7cf-b7ec7bb27821)

![image](https://camo.githubusercontent.com/2f4c4ec7fbde2135b7b7642c3d877c7169193ef1930cccc116241d278897d28d/68747470733a2f2f7668732e636861726d2e73682f7668732d353030786c34726d6a7a43624b6469724b6e7a4d446b2e676966)

joshka commented:
- Updated the output to point at `target/xxx.gif` to avoid having to ignore the files.
- Updated the gauge example to have a border around the last gauge (it doesn't address the issue, just makes it look a little nicer)
- Fixed the table example
- Updated the README
- Added instructions for updating the examples in a comment in the README

orhun commented:
@joshka Should I add the `vhs-action` in this PR or a separate one? Which one do you prefer?

joshka commented:
>@joshka Should I add the vhs-action in this PR or a separate one? Which one do you prefer?

Let's hold off on that for now. It only handles the generate part, not the publishing part, so if we wanted to fully automate the updating of the readme that might be a little difficult.

joshka commented:
- Braille dots are fixed in VHS [0.6.0](https://github.com/charmbracelet/vhs/releases/tag/v0.6.0). 
  - Removed notes about the dots in the README
  - Regenerated all images
- Updated calendar.tape to wait a little longer as the sleep was catching the last of the cargo build command

kdheepak commented on examples/user_input.tape:
```diff
@@ -0,0 +1,21 @@
+Sleep 1s
+Type "Goodbye, world!"
+Enter
+Sleep 5s
```
Should we add a lint check for checking if a new line exists at the end of a file?

kdheepak replied:
I tend to use `precommit` hooks locally and there's a `end-of-file-fixer`. I can look for something that we can run in CI if that's of interest.

joshka replied:
Yeah, go for it :)

kdheepak commented on examples/README.md:
```diff
@@ -0,0 +1,204 @@
+
+## Gauge ([gauge.rs](./gauge.rs))
+
+TODO: work out why the gauge background is rendering poorly via VHS.
```
What do you think about making these `TODO`s comments?

```suggestion
<!-- TODO: work out why the gauge background is rendering poorly via VHS.-->
```

joshka replied:
I wanted to make this visible so that users viewing the README would understand that this is not the how this should render and it's a problem with VHS not Ratatui. I've raised a bug with VHS at https://github.com/charmbracelet/vhs/issues/344 for this

kdheepak replied:
I see. Maybe we should change it to "the background is rendering poorly when we automate generating these examples, it will render correctly when you run this example." or something like that?

joshka replied:
Updated the messages

kdheepak commented on examples/README.md:
````diff
@@ -0,0 +1,209 @@
+
+![User Input][user_input.gif]
+
+<!--
+links to images to make it easier to update in bulk
+These are generated with `vhs publish examples/xxx.gif`
+
+To update these examples in bulk:
+```shell
+# build to ensure that running the examples doesn't have to wait so long
+cargo build --examples --features=crossterm,all-widgets
+for i in examples/*.tape
+do
+    echo -n "[${i:s:examples/:::s:.tape:.gif:}]: "
+    vhs $i --publish --quiet
+    # may need to adjust this depending on if you see rate limiting from VHS
+    sleep 1
+done
+```
+-->
````
I think it would be nice for us to have a developer documentation somewhere, either in this repo or part of the ratatui book.

joshka replied:
Right now that's https://github.com/ratatui-org/ratatui/blob/main/CONTRIBUTING.md, but we could split some of that out to DEVELOPING.md (or whatever the convention is)
